wxGUI: fix SearchCtrl in startup location wizard

nilason opened the pull request:
This should fix the search box issues in Startup > Create Location.

Addresses two failing code parts:
- use wx.EVT_SEARCHCTRL_SEARCH_BTN or wx.EVT_SEARCH to handle SearchCtrl events
- make sure ItemList data is a list

Tested on mac. wxpython 4.1 and 4.0.7
Needs testing on windows and linux.

Fixes #1319

nilason commented:
Note: this is 7.8 release branch only.

nilason commented:
@cmbarton if you want to try this, the easiest way is to replace:

gui/wxpython/location_wizard/wizard.py

with:
[wizard.py](https://raw.githubusercontent.com/OSGeo/grass/26ab13b3fec2dbed6918c6d0aea148f13d44c5b8/gui/wxpython/location_wizard/wizard.py)

On a 7.8.6dev or possibly 7.8.5.

nilason commented:
@petrasovaa now I see this is still on master. If this works as expected this should go there too, no?

cmbarton commented:
I'll test it in a few minutes

petrasovaa commented on gui/wxpython/location_wizard/wizard.py:
```diff
@@ -482,7 +487,10 @@ def __init__(self, wizard, parent):
 
         # projection list
-        self.projlist = ItemList(self, data=self.parent.projdesc.items(),
+        data = list()
+        for item in self.parent.projdesc.items():
+            data.append(item)
+        self.projlist = ItemList(self, data=data,
```
why not shorter `list(self.parent.projdesc.items())` ?

nilason replied:
Good point, fixed.

petrasovaa commented on gui/wxpython/location_wizard/wizard.py:
```diff
@@ -72,6 +72,11 @@
 global wizerror
 global translist
 
+if globalvar.CheckWxVersion(version=[4, 1, 0]):
+    search_evt = wx.EVT_SEARCH
+else:
+    search_evt = wx.EVT_SEARCHCTRL_SEARCH_BTN
+
```
This event is emitted when pressing the search button, right? Why not use (instead or in addition to) wx.EVT_TEXT to search as you type? That seems more intuitive.

nilason replied:
No button involved. Works on Mac with pressing enter with search field in focus. Just like it previously worked in Windows, where it actually worked. The question is whether it works with this fix in Linux and Windows.

nilason replied:
With second update commit wx.EVT_TEXT is used.

petrasovaa commented on gui/wxpython/location_wizard/wizard.py:
```diff
@@ -559,7 +567,7 @@ def OnSearch(self, event):
         try:
             self.proj, self.projdesc = self.projlist.Search(
-                index=[0, 1], pattern=event.GetString())
+                index=[0, 1], pattern=str)
```
Since you are editing this, it may be better to use different variable name instead of str, which is builtin function.

nilason replied:
Fixed here and another place with similar usage of `str`.

nilason commented:
Realised, there was no way to cancel or reset a search, so I added this with latest commit.
I also addressed wx.ListItemAttr() which was deprecated with wxpython 4.1.

nilason commented:
With latest commit I used wx.EVT_TEXT instead of wx.EVT_SEARCH. The advantage being dynamic update of filter/search, compared to pressing enter for initialising the search.

petrasovaa commented on gui/wxpython/location_wizard/wizard.py:
```diff
@@ -603,9 +614,13 @@ def __init__(self, parent, columns, data=None):
         # add some attributes
         #
-        self.attr1 = wx.ListItemAttr()
+        if globalvar.CheckWxVersion(version=[4, 1, 0]):
+            self.attr1 = wx.ItemAttr()
+            self.attr2 = wx.ItemAttr()
+        else:
+            self.attr1 = wx.ListItemAttr()
+            self.attr2 = wx.ListItemAttr()
         self.attr1.SetBackgroundColour(wx.Colour(238, 238, 238))
```
There seems to be other places with (List)ItemAttr, could you put it instead in gui_core/wrap.py to minimize duplication, something like:

```
if globalvar.CheckWxVersion(version=[4, 1, 0]):
    from wx import ItemAttr
else:
    from wx import ListItemAttr as ItemAttr 
```
(might need to use # noqa: F401  for flake checks)

nilason replied:
Realised it must be preferable to let the system set alternate row colour.
Used `ListCtrl.self.EnableAlternateRowColours()` instead of hard coding with latest commit.

nilason commented:
Tested now also current PR with 7.8.5 on Windows, works as expected.

nilason commented:
@petrasovaa Thank you!

petrasovaa commented:
This is at least partially relevant for master as well, right?

nilason commented:
Yes, working on it now.

nilason commented:
Now also pushed to master.
